File: moderate/sameend.py
#!/usr/bin/env python
# encoding: utf-8
"""
sameend.py

Created by Yamato Matsuoka on 2012-07-16.

Description:

You are given two strings 'A' and 'B'. Print out a 1 if string 'B' occurs at the end of string 'A'. Else a zero.

Input sample:

The first argument is a file, containing two strings, comma delimited, one per line. Ignore all empty lines in the input file.e.g. 

Hello World,World
Hello CodeEval,CodeEval
San Francisco,San Jose

Output sample:

Print out 1 if the second string occurs at the end of the first string. Else zero. Do NOT print out empty lines between your output. 
e.g.

1
1
0

"""

import sys

# def is_sameend(entry):
#     """Return 1 if the last word of the first string is the same 
#     as the first word of the second string. 0 otherwise."""
#     first, second = entry
#     return first.split()[-1] == second.split()[0]

def is_sameend(entry):
    """Return 1 if the last word of the first string is the same 
    as the first word of the second string. 0 otherwise."""
    first, second = entry
    lenB = len(second)
    if len(first) < lenB:
        return False
    else:
        return first[-lenB:] == second


if __name__ == '__main__':
    with open(sys.argv[1], "r") as f:
        data = [[s for s in line.rstrip().split(",")] for line in f if line.rstrip()]
    
    out = (int(is_sameend(x)) for x in data)
    print "\n".join(str(n) for n in out)

File: moderate/sameend.hs
{-
sameend.hs

Created by Yamato Matsuoka on 2012-07-09.

Description
------------
You are given two strings 'A' and 'B'. Print out a 1 if string 'B' occurs at the end of string 'A'. Else a zero.

Input sample
-------------
The first argument is a file, containing two strings, comma delimited, one per line. Ignore all empty lines in the input file.e.g. 
```
Hello World,World
Hello CodeEval,CodeEval
San Francisco,San Jose
```

Output sample
--------------
Print out 1 if the second string occurs at the end of the first string. Else zero. Do NOT print out empty lines between your output. e.g.
```
1
1
0
```
-}

import System.Environment (getArgs)

isSameEnd :: String -> String -> Bool
isSameEnd former latter
    | len1 < len2 = False
    | otherwise   = drop (len1 - len2) former == latter
       where 
         len1 = length former
         len2 = length latter

boolToInt :: Bool -> Int 
boolToInt True  = 1
boolToInt False = 0

parser :: String -> (String, String)
parser s = (former, tail latter)
    where (former, latter) = break (== ',') s

main = do 
    args <- getArgs
    contents <- readFile (head args)
    let inputs = map parser $ filter (not . null) $ lines contents
    let outputs = [boolToInt (isSameEnd s1 s2) | (s1, s2) <- inputs]
    mapM print outputs

